src/annotator/anchoring/pdf.js

```diff
@@ -0,0 +1,430 @@
+    return textContent;
+  };
+
+  // FIXME - `pdfViewer.getPageTextContent` was removed in recent versions of PDF.js.
+  pageTextCache[pageIndex] = PDFViewerApplication.pdfViewer
+    .getPageTextContent(pageIndex)
+    .then(joinItems);
+
+  return pageTextCache[pageIndex];
+}
+
+/**
+ * Return the offset in the text for the whole document at which the text for
+ * `pageIndex` begins.
+ *
+ * @param {number} pageIndex
+ * @return {Promise<number>}
+ */
+function getPageOffset(pageIndex) {
+  let index = -1;
+
+  const next = offset => {
+    ++index;
+    if (index === pageIndex) {
+      return Promise.resolve(offset);
+    }
+
+    return getPageTextContent(index).then(textContent =>
+      next(offset + textContent.length)
+    );
+  };
+
+  return next(0);
+}
+
+/**
+ * @typedef PageOffset
+ * @prop {number} index - Page index
+ * @prop {number} offset - Offset of the start of the text within the page
+ * @prop {string} textContent - Complete text of the page
+ */
+
+/**
+ * Convert an offset in the text of a document into an offset of the text within
+ * a page.
+ *
+ * @param {number} offset
+ * @return {PageOffset}
+ */
+function findPage(offset) {
+  let index = 0;
+  let total = 0;
+
+  // We call `count` once for each page, in order. The passed offset is found on
+  // the first page where the cumulative length of the text content exceeds the
+  // offset value.
+  //
+  // When we find the page the offset is on, we return an object containing the
+  // page index, the offset at the start of that page, and the textContent of
+  // that page.
+  //
+  // To understand this a little better, here's a worked example. Imagine a
+  // document with the following page lengths:
+  //
+  //    Page 0 has length 100
+  //    Page 1 has length 50
+  //    Page 2 has length 50
+  //
+  // Then here are the pages that various offsets are found on:
+  //
+  //    offset | index
+  //    --------------
+  //    0      | 0
+  //    99     | 0
+  //    100    | 1
+  //    101    | 1
+  //    149    | 1
+  //    150    | 2
+  const count = textContent => {
+    const lastPageIndex = PDFViewerApplication.pdfViewer.pagesCount - 1;
+    if (total + textContent.length > offset || index === lastPageIndex) {
+      offset = total;
+      return Promise.resolve({ index, offset, textContent });
+    } else {
+      ++index;
+      total += textContent.length;
+      return getPageTextContent(index).then(count);
+    }
+  };
+
+  return getPageTextContent(0).then(count);
+}
+
+/**
+ * Map a position anchor to a region within the text layer of a rendered page.
+ *
+ * If the page is not rendered, a placeholder element is created and the
+ * returned range is the placeholder.
+ *
+ * @param {PDFPageView} page - The PDF.js viewer page
+ * @param {TextPositionAnchor} anchor - Anchor to locate in page
+ * @param {Object} options - Options for `anchor.toSelector`
+ * @return {Range}
+ */
+function anchorByPosition(page, anchor, options) {
```

Comment:
It's right at this function that the cognitive load of this module gets uncomfortable.